src/meta-srv/src/service/store/cached_kv.rs

```diff
@@ -0,0 +1,472 @@
+    store: KvStoreRef,
+    cache: ResettableKvStoreRef,
+    // To avoid cache corruption, a version number is assigned each time the cache is updated.
+    // If the version number does not match, the related cache will be considered invalid until
+    // it is reloaded.
+    version: AtomicUsize,
+}
+
+impl LeaderCachedKvStore {
+    pub fn new(check_leader: CheckLeaderRef, store: KvStoreRef) -> Self {
+        Self {
+            check_leader,
+            store,
+            cache: Arc::new(MemStore::new()),
+            version: AtomicUsize::new(0),
+        }
+    }
+
+    /// With a leader checker which always returns true when checking,
+    /// mainly used in test scenarios.
+    pub fn with_always_leader(store: KvStoreRef) -> Self {
+        Self::new(Arc::new(AlwaysLeader), store)
+    }
+
+    #[inline]
+    fn is_leader(&self) -> bool {
+        self.check_leader.check()
+    }
+
+    #[inline]
+    async fn invalid_key(&self, key: Vec<u8>) -> Result<()> {
+        let _ = self.cache.delete(key, false).await?;
+        Ok(())
+    }
+
+    #[inline]
+    async fn invalid_keys(&self, keys: Vec<Vec<u8>>) -> Result<()> {
+        let txn = Txn::new().and_then(keys.into_iter().map(TxnOp::Delete).collect::<Vec<_>>());
+        let _ = self.cache.txn(txn).await?;
+        Ok(())
+    }
+
+    #[inline]
+    fn create_new_version(&self) -> usize {
+        self.version.fetch_add(1, Ordering::Relaxed) + 1
+    }
+
+    #[inline]
+    fn validate_version(&self, version: usize) -> bool {
+        version == self.version.load(Ordering::Relaxed)
+    }
+}
+
+#[async_trait::async_trait]
+impl KvStore for LeaderCachedKvStore {
+    async fn range(&self, req: RangeRequest) -> Result<RangeResponse> {
+        if !self.is_leader() {
+            return self.store.range(req).await;
+        }
+
+        // We can only cache for exact key queries (i.e. get requests)
+        // because we cannot confirm if a range response is complete.
+        if !req.range_end.is_empty() {
+            return self.store.range(req).await;
+        }
+
+        let res = self.cache.range(req.clone()).await?;
+        if !res.kvs.is_empty() {
+            return Ok(res);
+        }
+
+        let res = self.store.range(req.clone()).await?;
+        if !res.kvs.is_empty() {
+            let ver = self.create_new_version();
+
+            let KeyValue { key, value } = res.kvs[0].clone();
+            let put_req = PutRequest {
+                key: key.clone(),
+                value,
+                ..Default::default()
+            };
+            let _ = self.cache.put(put_req).await?;
+
+            if !self.validate_version(ver) {
+                self.invalid_key(key).await?;
+            }
+        }
+
+        return Ok(res);
+    }
+
+    async fn put(&self, req: PutRequest) -> Result<PutResponse> {
+        if !self.is_leader() {
+            return self.store.put(req).await;
+        }
+
+        let ver = self.create_new_version();
```

Comment:
Is it safe to put this line after the `let res = self.store.put(req.clone()).await?;`